### PR TITLE
Adding list of rules in ciliumspec

### PR DESCRIPTION
**Pull request by aanm**
Fixes https://github.com/cilium/cilium/issues/900
Signed-off-by: André Martins <andre@cilium.io>

### CODE REVIEW

**Comment by tgraf** [comment]
What are we missing here before this can get merged?

**Comment by aanm** [comment]
@tgraf I'm currently writing runtime tests. If you want we can merge and I'll push the tests later

**Comment by tgraf** [comment]
> @tgraf I'm currently writing runtime tests. If you want we can merge and I'll push the tests later

I'm fine with waiting until the runtime tests are done. I was just wondering if this needed more review first.

**Comment by aanm** [comment]
@tgraf I forgot to ask you in the meeting, on a cilium multi rule policy, which labels should I assign to each policy imported to cilium? Can we have 2+ policies with the same labels?

File: tests/k8s/multi-node/tests/02-cnp-specs.sh [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+kubectl create -f "${bookinfo_dir}/policies"
+
+if [ $? -ne 0 ]; then abort "policies were not inserted in kubernetes" ; fi
+
+cilium_id=$(docker ps -aq --filter=name=cilium-agent)
+
+docker exec -i ${cilium_id} cilium policy get io.cilium.k8s-policy-name=multi-rules 1>/dev/null
+
+# FIXME we have 2 policies with the same labels, figure it out how to deal
```
Comment (aanm) [inline tests/k8s/multi-node/tests/02-cnp-specs.sh]: @tgraf ^^

File: daemon/k8s_watcher.go [hunk 2]
```diff
@@ -835,9 +835,6 @@ func (d *Daemon) addCiliumNetworkPolicy(obj interface{}) {
 		return
 	}
 
-	// Delete an eventual existing rule with matching label
```
Comment (tgraf) [inline daemon/k8s_watcher.go:838]: Why are you removing this?

File: tests/k8s/multi-node/tests/02-cnp-specs.sh (hunk 1 above)
Comment (tgraf) [reply tests/k8s/multi-node/tests/02-cnp-specs.sh]: The default behavior on matching labels is to replace. So we will never have multiple rules with identical labels.
We could also enforce this in `Parse()` and search if multiple rules have identical labels and error out. Both is fine I think.

**Review by tgraf — CHANGES_REQUESTED** [review]

**Review by michi-covalent — APPROVED** [review]
will i be able to replace existing rules with the same name with this patch, or do i need to delete the existing rules first and then create new rules with the same name?

File: daemon/k8s_watcher.go (hunk 2 above)
Comment (aanm) [reply daemon/k8s_watcher.go:838]: I thought it was a leftover, do we need it? If yes, doesn't the `opts := AddOptions{Replace: true}` do the same?

Comment (tgraf) [reply daemon/k8s_watcher.go:838]: It indeed looks like a leftover. Removing this is correct.

**Comment by tgraf** [comment]
> will i be able to replace existing rules with the same name with this patch, or do i need to delete the existing rules first and then create new rules with the same name?

Each rule has a set of labels and will overwrite *all* rules with the same set of labels. If adding fails, *all* old rules will be put back in place.

**Comment by tgraf** [comment]
> @tgraf I forgot to ask you in the meeting, on a cilium multi rule policy, which labels should I assign to each policy imported to cilium? Can we have 2+ policies with the same labels?

Multiple rules with identical labels is actually just fine. All of them will be replaced.

**Comment by michi-covalent** [comment]
> Each rule has a set of labels and will overwrite all rules with the same set of labels. If adding fails, all old rules will be put back in place.

What happens if do something like this? First, create a policy with two rules:
```
apiVersion: "cilium.io/v1"
kind: CiliumNetworkPolicy
description: "Policy to test multiple rules in a single file"
metadata:
  name: "multi-rules"
specs:
  - endpointSelector:
      matchLabels:
        app: ratings
    ingress:
    - fromEndpoints:
      - matchLabels:
          app: reviews
  - endpointSelector:
      matchLabels:
        app: details
    ingress:
    - fromEndpoints:
      - matchLabels:
          app: productpage
```
Then, remove one rule from the YAML file and apply it again:
```
apiVersion: "cilium.io/v1"
kind: CiliumNetworkPolicy
description: "Policy to test multiple rules in a single file"
metadata:
  name: "multi-rules"
specs:
  - endpointSelector:
      matchLabels:
        app: ratings
    ingress:
    - fromEndpoints:
      - matchLabels:
          app: reviews
```
Will this remove the ingress rule from productpage to details?

**Comment by tgraf** [comment]
> Will this remove the ingress rule from productpage to details?

Yes. Identification of rules is based on the metadata and not on the rules themselves. Your example does not contain any labels in the rules themselves which means that the only label attached to the rules is the one which is generated from the policy name. Your delete request will thus delete both rules.

**Comment by michi-covalent** [comment]
Cool thanks @tgraf. One more follow up question. Say I created a policy with metadata like this:

```
metadata:
  name: my-policy-with-a-bunch-of-rules
  namespace: default
  labels:
    policy-version: v1
```
and later on I want to replace it with a new version of the policy:
```
metadata:
  name: my-policy-with-a-bunch-of-rules
  namespace: default
  labels:
    policy-version: v2
```
but I guess this won't match the existing policy because policy-version is different?

**Comment by aanm** [comment]
@tgraf I add a new commit as a workaround, it should be reverted once the l7 multinode bug is fixed.

**Review by tgraf — APPROVED** [review]

**Pull request merged**